### DISCUSSION TITLE
Naive implementation of transactions in 2.0.0

### DISCUSSION THREAD

**Discussion by lefam** [prisma/prisma#6734] [Q&A]
Hi,

I am running prisma version 2.0.0, and I can't easily upgrade right now. So, I am not able to leverage the transaction API available in recent versions.

What do you think about this naive implementation?

```
private async wrapTransaction<T>(codeBlock: () => Promise<T>) {
  await prisma.executeRaw('START TRANSACTION')
  try {
    const ret = await codeBlock()
    await prisma.executeRaw('COMMIT')
    return ret
  } catch (e) {
    console.error('Transaction failed: ', e)
    await prisma.executeRaw('ROLLBACK')
  }
}
```

Sample usage:

```
await wrapTransaction(async () => {
  // mix of db and non-db potentially async stuff
});
```

What do you think of the implementation above? What can go wrong?

Thanks in advance.

**Comment by janpio** [comment]
Each `prisma` call can under the hood be executed by a different database connection - so your `ROLLBACK` might be on a connection that did not start the transaction, or did not execute the actual query, or both.

You might think that limiting the connection pool size to 1 (`connection_limit=1`) might help, and it certainly would make it better (more probable) - but in theory the connection could also be exchanged (old one closed, opened new one) between these queries.

This would unfortunately not give you the guarantees you are probably hoping for when using transactions.

**Reply by janpio** [reply]
Colleague pointed out, that the `connection_limit=1` idea makes even less sense if you have multiple requests coming in to your project (which we all hope you do) as that could totally send a query in your transaction in this case! Or create a second one, commit the first one before you did the actual query and so on.

So: Unfortunately does not work at all in any reasonable way.
